libs/mesh/src/meshCubatureSetupQuad2D.cpp

```diff
@@ -72,47 +71,46 @@ void meshQuad2D::CubatureSetup(){
   props["defines/" "p_intNfpNfaces"]= intNfp*Nfaces;
   props["defines/" "p_cubNfp"]= cubNfp;
 
-  dfloat *cubDWT = (dfloat*) calloc(cubNq*Nq, sizeof(dfloat));
+  // build transposes (we hold matrices as column major on device)
   dfloat *cubProjectT = (dfloat*) calloc(cubNq*Nq, sizeof(dfloat));
-  dfloat *cubInterpT = (dfloat*) calloc(cubNq*Nq, sizeof(dfloat));
-  for(int n=0;n<Nq;++n){
-    for(int m=0;m<cubNq;++m){
-      cubDWT[n+m*Nq] = cubDW[n*cubNq+m];
-      cubProjectT[n+m*Nq] = cubProject[n*cubNq+m];
-      cubInterpT[m+n*cubNq] = cubInterp[m*Nq+n];
+  dfloat *cubInterpT   = (dfloat*) calloc(cubNq*Nq, sizeof(dfloat));
+  matrixTranspose(cubNq, Nq, cubInterp, Nq, cubInterpT, cubNq);
+  matrixTranspose(Nq, cubNq, cubProject, cubNq, cubProjectT, Nq);
+
+  //pre-multiply cubProject by W on device
+  for(int n=0;n<cubNq;++n){
+    for(int m=0;m<Nq;++m){
+      cubProjectT[m+n*Nq] *= cubw[n];
     }
   }
 
-  o_cubInterpT =
-    device.malloc(Nq*cubNq*sizeof(dfloat),
-        cubInterpT);
+  dfloat *cubPDTT     = (dfloat*) calloc(cubNq*Nq, sizeof(dfloat));
+  matrixTranspose(Nq, cubNq, cubPDT, cubNq, cubPDTT, Nq);
 
-  o_cubProjectT =
-    device.malloc(Nq*cubNq*sizeof(dfloat),
-        cubProjectT);
-
-  o_cubDWT =
-    device.malloc(Nq*cubNq*sizeof(dfloat),
-        cubDWT);
+  //pre-multiply cubPDT by W on device
+  for(int n=0;n<cubNq;++n){
+    for(int m=0;m<Nq;++m){
+      cubPDTT[m+n*Nq] *= cubw[n];
+    }
+  }
 
-  o_cubDWmatrices = device.malloc(cubNq*Nq*sizeof(dfloat), cubDWT);
+  o_cubInterp   = device.malloc(Nq*cubNq*sizeof(dfloat), cubInterpT);
+  o_cubProject = device.malloc(Nq*cubNq*sizeof(dfloat), cubProjectT);
 
-  o_intInterpT = device.malloc(cubNq*Nq*sizeof(dfloat));
-  o_intInterpT.copyFrom(o_cubInterpT);
+  o_cubPDT = device.malloc(Nq*cubNq*sizeof(dfloat), cubPDTT);
+  o_cubD = device.malloc(cubNq*Nq*sizeof(dfloat), cubD);
```

Comment:
The size of `o_cubD` here should be `cubNq*cubNq`.

Reply:
Good call. I fixed this in #38 after I ran into a bug it caused, but I'll patch it in this PR for completeness.

Reply:
Scratch that, I didn't notice this was quads. I'll merge the fix with I rebase #38.